src/MongoDB.Driver/Linq/Translators/AggregateLanguageTranslator.cs

```diff
@@ -1276,5 +1303,57 @@ private bool TryTranslateSumResultOperator(PipelineExpression node, out BsonValu
             result = null;
             return false;
         }
+
+        private static class TemplateHelper
+        {
+            private const string MapTemplateStartDelimiter = "#{{";
+            private const string MapTemplateEndDelimiter = "}}#";
+            private const string MapTemplate = MapTemplateStartDelimiter + "{{{0}}}" + MapTemplateEndDelimiter;
+
+            private static string GetNotReplacedTemplate(string value)
+            {
+                var startIndexOf = value.IndexOf(MapTemplateStartDelimiter, StringComparison.Ordinal);
+                var endIndexOf = value.IndexOf(MapTemplateEndDelimiter, StringComparison.Ordinal);
+                if (startIndexOf != -1 && endIndexOf != -1 && endIndexOf > startIndexOf)
+                {
+                    endIndexOf = endIndexOf + MapTemplateEndDelimiter.Length;
+                    if (endIndexOf <= value.Length) return value.Substring(startIndexOf, endIndexOf - startIndexOf);
+                }
+
+                return null;
+            }
+
+            public static BsonValue ReplaceTemplates(BsonValue bsonValue, Dictionary<string,string> mapHistory)
+            {
+                var value = bsonValue.ToString();
+                foreach (var map in mapHistory)
+                {
+                    value = value.Replace(string.Format(MapTemplate, map.Value), map.Key);
+                }
+
+                var topLevelParameter = GetNotReplacedTemplate(value);
+                if (!string.IsNullOrWhiteSpace(topLevelParameter) && mapHistory.Any())
+                {
+                    // This case can happen no more than once when we use a top-level input expression parameter
+                    // in the scope of children expressions. Since we don't have any mentioning of the top-level input parameters
+                    // in any SERVER operators, then we can have the single not replaced template 
+                    // which will mean that this is a top-level input parameter.
+                    // We need to handle this case additionally.
+                    value = value.Replace(topLevelParameter, mapHistory.Last().Key);
+                }
+
+                if (value.Contains(MapTemplateStartDelimiter) || value.Contains(MapTemplateEndDelimiter))
+                {
+                    throw new MongoException("The query transformation has not been finished: Incorrect parameters mapping.");
+                }
+
+                return BsonDocument.Parse(value);
```

Comment:
How do you know value is a parseable BsonDocument? It was only a BsonValue coming in.

Reply:
Agree. I will replace this on `BsonValue.Create(value)`.

Reply:
BsonValue.Create doesn't sound right either. Since value is a string BsonValue.Create will just return a BsonString.

Reply:
I've changed this code.